pczt/src/orchard.rs

```diff
@@ -0,0 +1,536 @@
+    /// the values and change involved in the transaction.
+    ///
+    /// This exposes the value to all participants. For Signers who don't need this
+    /// information, we can drop the values and compress the rcvs into the bsk global.
+    pub(crate) value: Option<u64>,
+
+    /// The seed randomness for the output.
+    ///
+    /// - This is set by the Constructor.
+    /// - This is required by the Prover, instead of disclosing `shared_secret` to them.
+    #[getset(get = "pub")]
+    pub(crate) rseed: Option<[u8; 32]>,
+
+    /// The `ock` value used to encrypt `out_ciphertext`.
+    ///
+    /// This enables Signers to verify that `out_ciphertext` is correctly encrypted.
+    ///
+    /// This may be `None` if the Constructor added the output using an OVK policy of
+    /// "None", to make the output unrecoverable from the chain by the sender.
+    pub(crate) ock: Option<[u8; 32]>,
+
+    /// The ZIP 32 derivation path at which the spending key can be found for the output.
+    pub(crate) zip32_derivation: Option<Zip32Derivation>,
+
+    /// Proprietary fields related to the note being created.
+    #[getset(get_mut = "pub")]
+    pub(crate) proprietary: BTreeMap<String, Vec<u8>>,
+}
+
+impl Bundle {
+    /// Merges this bundle with another.
+    ///
+    /// Returns `None` if the bundles have conflicting data.
+    pub(crate) fn merge(mut self, other: Self) -> Option<Self> {
+        // Destructure `other` to ensure we handle everything.
+        let Self {
+            mut actions,
+            flags,
+            value_sum,
+            anchor,
+            zkproof,
+            bsk,
+        } = other;
+
+        if self.flags != flags {
+            return None;
+        }
+
+        // If `bsk` is set on either bundle, the IO Finalizer has run, which means we
+        // cannot have differing numbers of actions, and the value sums must match.
+        match (self.bsk.as_mut(), bsk) {
+            (Some(lhs), Some(rhs)) if lhs != &rhs => return None,
+            (Some(_), _) | (_, Some(_))
+                if self.actions.len() != actions.len() || self.value_sum != value_sum =>
+            {
+                return None
+            }
+            // IO Finalizer has run, and neither bundle has excess spends or outputs.
+            (Some(_), _) | (_, Some(_)) => (),
+            // IO Finalizer has not run on either bundle. If the other bundle has more
+            // spends or outputs than us, move them over; these cannot conflict by
+            // construction.
+            (None, None) => {
+                if actions.len() > self.actions.len() {
+                    self.actions.extend(actions.drain(self.actions.len()..));
+
+                    // We check below that the overlapping actions match. Assuming here
+                    // that they will, we can take the other bundle's value sum.
+                    self.value_sum = value_sum;
+                }
+            }
+        }
+
+        if self.anchor != anchor {
```

Comment:
```suggestion
        if !merge_optional(&mut self.anchor, anchor) {
```
(if `anchor` is made optional).